### PR TITLE
RFC: Composable input/output pipeline

### PR DESCRIPTION
**Pull request by adriangb**
This is an attempt to collect ideas from various issues / PRs into a coherent framework and generalize our current input/output transformers.

@stsievert  would you mind taking a look?

### CODE REVIEW

File: rfcs/1-input-pipeline.md [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+class KerasRegressor:
+    def __init__(
+        self,
+        array_transformers = (RegressorInputValidator,)
+    ):
+        ...
+```
+
+## Issues this can potentially resolve
```
Comment (stsievert) [inline rfcs/1-input-pipeline.md:377]: These issues seem to be around modular validation/transforming. This RFC proposes one solution. What are other solutions? Why does this RFC represent the best solution?

File: rfcs/1-input-pipeline.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
+## Background
+
+One of the primary functions of the Scikit-Learn wrappers for Keras is fielding conversion from Scikit-Learn's data types
+to TensorFlow's data types. An example of these conversions is integer-encoding or one-hot encoding categorical targets.
+
+Originally, these conversions were hardcoded into the wrappers in an ad-hoc manner (see here).
```
Comment (stsievert) [inline rfcs/1-input-pipeline.md:6]: > ad-hoc manner (see here)

Link missing.

**Review by stsievert — COMMENTED** [review]
Thanks for this RFC – this is easier to review than a PR. Here are some questions:

* What issue are you trying to solve with this RFC? What's an illustration of that issue?
* Is that issue user-facing?
* How does this RFC solve that (user-facing?) issue?

File: rfcs/1-input-pipeline.md (hunk 2 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md]: Thank you, added. It was this: https://github.com/keras-team/keras/blob/80ca7a7a06058076fde5ee36f8ac1f2ca30f8996/keras/wrappers/scikit_learn.py#L155-L157

File: rfcs/1-input-pipeline.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+## Proposal
+
+This proposal consists of 2 pipelines:
+1. A pipeline for preparing array-like data for a `tf.data.Dataset`. This would include, for example, integer-encoding object-dtype arrays (`tf.Tensors` can't hold objects).
+2. A pipeline for applying transformations to `tf.data.Dataset`s. For example, one-hot encoding targets for classifiers using the categorical crossentropy loss.
+
+If the data comes in as a `tf.data.Dataset`, the first pipeline is skipped. If not, it is run and the output is converted to a `Dataset`.
+The second pipeline is then run in all cases.
+
+These pipelines will consist of chained transformers implementing a Scikit-Learn-like interface, but without being restricted to the exact Scikit-Learn API.
```
[the remark this answers is not in the document]
Comment (adriangb) [reply rfcs/1-input-pipeline.md:44]: By "exact Scikit-Learn API" I mean roughly:
- having a `fit` method with the signature `fit(X, y=None) -> self`
- having a transform method with the signature `transform(X) -> X'`
- an `inverse_transform` method with the signature `inverse_transform(X') -> X`.
And no other methods (that are part of the API).
Also `inverse_transform(transform(X)) == X` is not a requirement of the API, but it certainly is the spirit of it (and how most sklearn transformers are implemented, when it makes sense to do so).

> And the proposed "Scikit-learn-like interface" is public, correct?

Yes, it would be public and consists of `ArrayTransformer` and `DatasetTransformer`.
These can be base classes, protocols/interfaces or just duck-typing (implementation detail).
The idea is that users could take our default data validation/transformations and mix/match with their own for use cases like multi-output models or multi-output class reweighting.

This proposed interface violates the sklearn interface in several ways:
- by passing `X`, `y` and `sample_weight` in `ArrayTransformer`
- by passing a `tf.data.Dataset` object in `DatasetTransformer`
- because these really aren't inverse transformations: the forward/input transformation transforms X & y, the output transformation transforms `y'` (y predictions) and/or `y` itself. So `inverse_transform(transform(X)) != X`.
- by allowing side effects to be applied to the `model` (this is an addition to the API, it doesn't strictly break the implementation, but it does break the spirit).

> How can this RFC be reworked to conform with the "exact Scikit-Learn API"?

I don't think it can be easily reworked: there are too many limitations in the sklearn transformer API ((and the authors have said as much themselves)[https://github.com/scikit-learn/scikit-learn/issues/3855#issuecomment-63213164]).

The good news is that the Scikit-Learn API is a strict subset of this API, so we could provide wrappers to convert the Scikit-Learn transformer API into this one (eg. by specifying what `X` is to your transformer and dispatching the methods).

But let me ask: why is it important to adhere to the Scikit-Learn API? I know there are good reasons, I just want understand which you are thinking of.
My thoughts are that if we can't interoperate directly (i.e. use `make_pipeline` and other facilities of sklearn), there is not much value in being semi-compatible. I also think there aren't that many things (beyond pipelines) that would be useful here. But this may be short-sighted, I'm open to other opinions.

**Comment by adriangb** [comment]
> Thanks for this RFC – this is easier to review than a PR.

Good to know it helps! And thank you for reviewing.

> What issue are you trying to solve with this RFC? What's an illustration of that issue?

This solves some user-facing issues, as well as allowing us to rework things internally to be cleaner and more flexible for users.

There are several different issues, but I think the clearest one is #160.

The general user request is to be able to use `tf.data.Dataset`. 

We could hack this together as is (by hardcoding bypasses to `BaseWrapper._validate_data` and feature/target encoders), but even then we would lose attributes like `n_features_in_` that are both part of our API and the Scikit-Learn API. This RFC solves this problem by making processing of `tf.data.Dataset` a first-class citizen, and even allowing us to move some of our validation/preprocessing to use `tf.data.Dataset` which, for example, would allow us to provide `n_features_in_` and `classes_` regardless of if the input was array-like or a `tf.data.Dataset`.

The other issues linked are generally around the theme of more validations (#106, #143) or figuring out how to clean up our current validations (#111, #209). These two issues are intertwined because:
1. Every validation added has the chance of rubbing up against edge cases. If these validations aren't modular (eg if they're buried 3 layers deep into private method calls), we might block a use case we can't think of right now just to provide a nice error message.
2. If we add more ad-hoc private methods that do validation/transformation (eg `BaseWrapper._validate_data`, `BaseWrapper._check_model_compatibility`) we end up with a mess of private methods where it's hard to understand the execution order, goal of the method, etc.

This RFC resolves this by providing a unified interface for these validations/transformations that is modular, composable and public. I envision it helping our users (since they can disable or add validation/transformation steps) as well as us (the developers) because we can organize our default validation/transformations better.

File: rfcs/1-input-pipeline.md (hunk 1 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md:377]: > What are other solutions?

This is, of course, a great question. I think general solutions would be variations of the same idea, perhaps less structured (eg. hardcoding that `tf.data.Dataset` inputs should skip `BaseWrapper._validate_data`). I'd have to think a bit to see if I can come up with any other structured approaches that might also solve the issues.

File: rfcs/1-input-pipeline.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (stsievert) [reply rfcs/1-input-pipeline.md:43]: This library has one job: to bring the Scikit-learn API to Keras. That means there needs to be a ***really good*** reason to break the Scikit-learn API.

That reason needs to consider alternative implementations, including the one that exactly follows the Scikit-learn API. I imagine some questions that need to be answered include the following:


* What affect does the implementation have on developers?
* What affect does the implementation have on the user?
* What specific issues make the implementation inappropriate?

File: rfcs/1-input-pipeline.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+ArrayLike = Union[Sequence, np.ndarray]
+
+
+class NotInitializedError(Exception):
+    ...
+
+
+class ArrayTransformer:
+
+    def set_model(self, model: "BaseWrapper") -> None:
```
Comment (stsievert) [inline rfcs/1-input-pipeline.md:66]: :-1:

Why not follow the Scikit-Learn API?

``` python
class ArrayTransformer(BaseEstimator):
    def __init__(self, model=None):
        self.model = model

    ...

a = ArrayTransformer(model=model) # option 0

a = ArrayTransformer().model = model # option 1

a = ArrayTransformer()
a.set_params(model=model) # option 2
```

Yes, TF uses `set_model`. Why should their boilerplate be followed?

File: rfcs/1-input-pipeline.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        self.model = model
+
+    def transform_input(self, X: ArrayLike, y: Optional[ArrayLike], sample_weight: Optional[ArrayLike], *, initialize: bool = True) -> Data:
+        return X, y, sample_weight
+
+    def transform_output(self, y_pred_proba: np.ndarray, y: Union[np.ndarray, None]) -> Tuple[np.ndarray, Union[np.ndarray, None]]:
+        return y_pred_proba, None
+
+    def get_meta(self) -> Dict[str, Any]:
```
Comment (stsievert) [inline rfcs/1-input-pipeline.md:75]: :-1:

Why not have an attribute `meta_`?

File: rfcs/1-input-pipeline.md [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+class DatasetTransformer:
+
+    def set_model(self, model: "BaseWrapper") -> None:
+        self.model = model
+
+    def transform_input(self, data: tf.data.Dataset, *, initialize: bool = True) -> tf.data.Dataset:
+        return data
+
+    def transform_output(self, y_pred_proba: np.ndarray, y: Union[np.ndarray, None]) -> Tuple[np.ndarray, Union[np.ndarray, None]]:
```
Comment (stsievert) [inline rfcs/1-input-pipeline.md:87]: Why not rename these functions `inverse_transform`?

File: rfcs/1-input-pipeline.md (hunk 1 above)
Comment (stsievert) [reply rfcs/1-input-pipeline.md:378]: Why not have two Scikit-learn transformers, one for validation and one to change the data?

File: rfcs/1-input-pipeline.md (hunk 5 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md:75]: The `get_meta` is stuck in there from older revision.

My current thoughts are to not have any interface for these meta parameters and instead let the implementers set attributes on `model` directly (or I guess we could have a `meta_` attribute on `BaseWrapper` and then do some `__getattr__` stuff to expose these attributes).

Sorry for having old stuff left in there.

File: rfcs/1-input-pipeline.md (hunk 4 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md:66]: This would require users to give us a class to initialize (this object is being passed to the constructor that creates `model`).

I guess we could use parameter routing to allow setting any other parameters. Does this look any better?

```python
class MyTransf:

    def __init__(self, model, other):
        self.other = other
    
    ...

est = BaseWrapper(..., pipeline_param=[MyTransf], pipeline_param__0__other=True)
# or
est = BaseWrapper(..., pipeline_param={"tfname": MyTransf}, pipeline_param__ tfname__other=True)  # relies on dict ordering to know what order to run pipeline in
# or
est = BaseWrapper(..., pipeline_param=[("tfname", MyTransf)], pipeline_param__ tfname__other=True)  # same as an sklearn pipeline using a tuple to set the name
```

An alternative would be to require users to bind any other parameters using `functools.partial` or something.

File: rfcs/1-input-pipeline.md (hunk 6 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md:87]: Because their signature and usage does not match `inverse_transform`'s signature. They wouldn't work in an sklearn `Pipeline`, nor would they be chainable with sklearn estimators.

File: rfcs/1-input-pipeline.md (hunk 1 above)
Comment (adriangb) [reply rfcs/1-input-pipeline.md:378]: We would need:
1. Interface for validation using array-like data
2. Interface for data changing using array-like data
3. Interface for validation using Dataset
4. Interface for data changing using array-like data

I think this is too many interfaces. The only difference between a "validation" and a "transformation" is that the transformation _needs_ to return the data, but the validation does not. So by having the validation return the data we can collapse those two concepts into one. Validation transformers simply inspect the data but do not modify it.

Comment (stsievert) [reply rfcs/1-input-pipeline.md:378]: What use case/issue motivates having separate classes for Dataset/array-like inputs? Why not collapse it into one class?

Why is 4 interfaces "too many"? Why is the current framework with `target_encoder_` and `feature_encoder_` not sufficient?

I'm asking questions to get answers encoded in the RFC (an RFC is a really good idea to encode these decisions).
